peer/abstractlist/list.go

```diff
@@ -0,0 +1,592 @@
+func NoShuffle() Option {
+	return optionFunc(func(options *options) {
+		options.noShuffle = true
+	})
+}
+
+// FailFast indicates that the peer list should not wait for peers to be added,
+// when choosing a peer.
+//
+// This option is particularly useful for proxies.
+func FailFast() Option {
+	return optionFunc(func(options *options) {
+		options.failFast = true
+	})
+}
+
+// Seed specifies the random seed to use for shuffling peers
+//
+// Defaults to approximately the process start time in nanoseconds.
+func Seed(seed int64) Option {
+	return optionFunc(func(options *options) {
+		options.seed = seed
+	})
+}
+
+// New creates a new peer list with an identifier chooser for available peers.
+func New(name string, transport peer.Transport, implementation Implementation, opts ...Option) *List {
+	options := defaultOptions
+	for _, o := range opts {
+		o.apply(&options)
+	}
+
+	return &List{
+		once:               lifecycle.NewOnce(),
+		name:               name,
+		peers:              make(map[string]*peerFacade, options.capacity),
+		offlinePeers:       make(map[string]peer.Identifier, options.capacity),
+		implementation:     implementation,
+		transport:          transport,
+		noShuffle:          options.noShuffle,
+		failFast:           options.failFast,
+		randSrc:            rand.NewSource(options.seed),
+		peerAvailableEvent: make(chan struct{}, 1),
+	}
+}
+
+// List is an abstract peer list, backed by an Implementation to
+// determine which peer to choose among available peers.
+// The abstract list manages available versus unavailable peers, intercepting
+// these notifications from the transport's concrete implementation of
+// peer.StatusPeer with the peer.Subscriber API.
+// The peer list will not choose an unavailable peer, prefering to block until
+// one becomes available.
+//
+// The list is a suitable basis for concrete implementations like round-robin.
+type List struct {
+	lock sync.RWMutex
+	once *lifecycle.Once
+
+	name string
+
+	peers              map[string]*peerFacade
+	offlinePeers       map[string]peer.Identifier
+	implementation     Implementation
+	peerAvailableEvent chan struct{}
+	transport          peer.Transport
+
+	noShuffle bool
+	failFast  bool
+	randSrc   rand.Source
+}
+
+// Name returns the name of the list.
+func (pl *List) Name() string { return pl.name }
+
+// Transport returns the underlying transport for retaining and releasing peers.
+func (pl *List) Transport() peer.Transport { return pl.transport }
+
+// Update applies the additions and removals of peer Identifiers to the list
+// it returns a multi-error result of every failure that happened without
+// circuit breaking due to failures.
+func (pl *List) Update(updates peer.ListUpdates) error {
+	if len(updates.Additions) == 0 && len(updates.Removals) == 0 {
+		return nil
+	}
+
+	pl.lock.Lock()
+	defer pl.lock.Unlock()
+
+	if !pl.once.IsRunning() {
+		return pl.updateOffline(updates)
```

Comment:
Yeah, I considered that. I decided to make the behavior of update orthogonal to lifecycle since the updater is unlikely to coordinate with the lifecycle. This approach preserves the same errors if the updates are issued out of order, regardless of lifecycle order.